3p/iframe-transport-client.js

```diff
@@ -33,61 +34,124 @@ export class IframeTransportClient {
-    this.client_.setSentinel(dev().assertString(
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setHostWindow(this.win_.parent);
+    this.iframeMessagingClient_.setSentinel(dev().assertString(
         tryParseJson(this.win_.name)['sentinel'],
         'Invalid/missing sentinel on iframe name attribute' + this.win_.name));
-    this.client_.makeRequest(
+    this.iframeMessagingClient_.makeRequest(
         MessageType.SEND_IFRAME_TRANSPORT_EVENTS,
         MessageType.IFRAME_TRANSPORT_EVENTS,
         eventData => {
           const events =
-              /**
-               * @type
-               * {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
-               */
-              (eventData['events']);
+          /**
+           * @type
+           *   {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
+           */
+          (eventData['events']);
           user().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              user().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              const context = this.contextFor_(event.creativeId);
+              context.dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+              e.message);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    const vendor = tryParseJson(this.win_.name)['type'];
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, vendor));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    this.win_ = win;
```

Comment:
js doc.  Note that you never actually use the win outside of the constructor so really no need to store.  Similarly you could create a base message object saving you the need to store creativeId & vendor (note that my solution assumes you don't want to modify response parameter given to sendResponseToCreative):

/** @private @const {!Object} */
this.baseMessage_ = {creativeId, vendor};

* @param {!Object<string,string>} response
sendResponseToCreative(response) {
   this.iframeMessagingClient_./*OK*/sendMessage(
   MessageType.IFRAME_TRANSPORT_RESPONSE,
       dict(Object.assign({}, response, this.baseMessage_)));

Reply:
The linter didn't like that:
```
/Users/jonkeller/workspace/amp-ad-exit-3p/amphtml/3p/iframe-transport-client.js
  161:9  error  calls to `dict` must have an Object Literal Expression as the first argument  dict-string-keys
```
I then tried:
```
    this.iframeMessagingClient_./*OK*/sendMessage(
        MessageType.IFRAME_TRANSPORT_RESPONSE,
        /* @type {!JsonObject} */
        (Object.assign({}, message, this.baseMessage_)));
```
and check-types seemed to just ignore the cast, and gave me:
```
3p/iframe-transport-client.js:161: ERROR - actual parameter 2 of IframeMessagingClient$$module$3p$iframe_messaging_client.prototype.sendMessage does not match formal parameter
found   : Object
required: (JsonObject|null|undefined)
        (Object.assign({}, message, this.baseMessage_)));
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Any other ideas?